packages/engine-server/src/workspace/utils.ts

```diff
@@ -57,4 +68,97 @@ export class WorkspaceUtils {
     }
     return noteHash !== note.contentHash;
   }
+
+  /**
+   * Generate url for given note or return `undefined` if no url is specified
+   * @param opts
+   *
+   */
+  static getNoteUrl(opts: {
+    config: DendronConfig;
+    note: NoteProps;
+    vault: DVault;
+    urlRoot?: string;
+    maybeNote?: NoteProps;
+    anchor?: string;
+  }) {
+    const { config, note, anchor, vault } = opts;
+    let { urlRoot, maybeNote } = opts;
+    const notePrefix = "notes";
+    if (_.isUndefined(maybeNote)) {
+      maybeNote = note;
+    }
+    /**
+     * set to true if index node, don't append id at the end
+     */
+    let isIndex: boolean = false;
+
+    if (vault.seed) {
+      if (config.seeds && config.seeds[vault.seed]) {
+        const maybeSite = config.seeds[vault.seed]?.site;
+        if (maybeSite) {
+          urlRoot = maybeSite.url;
+          if (!_.isUndefined(maybeNote)) {
+            // if custom index is set, match against that, otherwise `root` is default index
+            isIndex = maybeSite.index
+              ? maybeNote.fname === maybeSite.index
+              : DNodeUtils.isRoot(maybeNote);
+          }
+        }
+      }
+    }
+    let root = "";
+    if (!_.isUndefined(urlRoot)) {
+      root = urlRoot;
+    } else {
+      // assume github
+      throw Error("not implemented");
```

Comment:
convert to use `DendronError`.  What does "assume github" mean here?  If we decide to keep the `urlRoot` parameter, then it should be a mandatory parameter.

Reply:
After Refactoring copyNoteUrl command, the urlRoot can be undefined.
```ts
const urlRoot =
      config?.site?.siteUrl ||
      DendronWorkspace.configuration().get<string>(
        CONFIG.COPY_NOTE_URL_ROOT.key
      );
```

Reply:
In this case - if urlRoot is not defined, then let's go ahead and stop the command within copyNoteUrl itself.  There's no reason to continue if there's no base URL.

Reply:
If urlRoot is not defined here, we move forward and check if the vault is seed. If it is a seed, we determine the value of urlRoot from 
```ts
if (vault.seed) {
      if (config.seeds && config.seeds[vault.seed]) {
        const maybeSite = config.seeds[vault.seed]?.site;
        if (maybeSite) {
          urlRoot = maybeSite.url;
}
```

Else we stop and throw error.